controllers/PlagiarismIthenticateActionHandler.inc.php

```diff
@@ -0,0 +1,385 @@
+ *
+ * Copyright (c) 2014-2024 Simon Fraser University
+ * Copyright (c) 2000-2024 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file LICENSE.
+ *
+ * @class PlagiarismIthenticateActionHandler
+ * @ingroup plugins_generic_plagiarism
+ *
+ * @brief Handle the different iThenticate service related actions
+ */
+
+import("plugins.generic.plagiarism.controllers.PlagiarismComponentHandler");
+import('lib.pkp.classes.core.JSONMessage'); 
+import('lib.pkp.classes.db.DAORegistry');
+import('classes.notification.NotificationManager');
+import("plugins.generic.plagiarism.IThenticate");
+
+class PlagiarismIthenticateActionHandler extends PlagiarismComponentHandler {
+
+	/**
+	 * Launch the iThenticate similarity report viewer
+	 *
+	 * @param array $args
+	 * @param Request $request
+	 */
+	public function launchViewer($args, $request) {
+		$context = $request->getContext();
+		$user = $request->getUser();
+
+		$submissionFileDao = DAORegistry::getDAO('SubmissionFileDAO'); /** @var SubmissionFileDAO $submissionFileDao */
+		$submissionFile = $submissionFileDao->getById($args['submissionFileId']);
+		$submissionDao = DAORegistry::getDAO('SubmissionDAO'); /** @var SubmissionDAO $submissionDao */
+		$submission = $submissionDao->getById($submissionFile->getData('submissionId'));
+		$siteDao = DAORegistry::getDAO("SiteDAO"); /** @var SiteDAO $siteDao */
+		$site = $siteDao->getSite();
+
+		/** @var \IThenticate $ithenticate */
+		$ithenticate = static::$_plugin->initIthenticate(
+			...static::$_plugin->getServiceAccess($context)
+		);
+
+		$locale = $ithenticate
+			->setApplicableEulaVersion($submission->getData('ithenticateEulaVersion'))
+			->getApplicableLocale(
+				collect([$submission->getData("locale")])
+					->merge($user->getData("locales"))
+					->merge([$context->getPrimaryLocale(), $site->getPrimaryLocale()])
+					->unique()
+					->toArray()
+			);
+
+		$viewerUrl = $ithenticate->createViewerLaunchUrl(
+			$submissionFile->getData('ithenticateId'),
+			$user,
+			$locale,
+			static::$_plugin->getSubmitterPermission($context, $user),
+			(bool)static::$_plugin->getSimilarityConfigSettings($context, 'allowViewerUpdate')
+		);
+
+		if (!$viewerUrl) {
+			return $request->redirect(
+				null,
+				'user',
+				'authorizationDenied',
+				null,
+				['message' => 'plugins.generic.plagiarism.action.launchViewer.error']
+			);
+		}
+
+		return $request->redirectUrl($viewerUrl);
+	}
+
+	/**
+	 * Schedule the similarity report generate process at iThenticate services's end
+	 *
+	 * @param array $args
+	 * @param Request $request
+	 */
+    public function scheduleSimilarityReport($args, $request) {
+
+		$context = $request->getContext();
+
+		/** @var SubmissionFileDAO $submissionFileDao */
+		$submissionFileDao = DAORegistry::getDAO('SubmissionFileDAO');
+		$submissionFile = $submissionFileDao->getById($args['submissionFileId']);
```

Comment:
As above, this is dangerous -- any file ID in the system can be specified here and it'll be trusted by the code.